### DISCUSSION TITLE
Kanban basics (and how to implement it in GitHub)

### DISCUSSION THREAD

**Discussion by lakruzz** [thetechcollective/public-discussions#15] [Contemporary DevOps]
**In this post I'll share a few rule-of-thumbs for kanban task management[^taskmanagement] and for each of them elaborate a bit on a it's practical real-life implementation specifically in GitHub Projects.**

[^taskmanagement]: I deliberately do not call it _project management_ but rather _task management;_ this approach - kanban - applies to all kind of work, not just project oriented work. It's concerned with organizing tasks - in general.

# Kanban Rules Of Thumbs

👍   1: Organize your project so you have a) an **inbox**, b) a **backlog** and c) a **kanban board** of at least three columns; _ToDo_, _Work In Progress_ and _Done_
👍   2: Keep everything in your ToDo list prioritized - at all times
👍   3: Kanban is a _pull_ model as opposed to a _push_ model. Decide how exactly this _pull_ is implemented in your team
👍   4: Groom your kanban board at least once every day

## 👍 1: Organize your project in inbox, backlog and a kanban

The basic kanban board is just three columns: `ToDo`, `Work In Progress` (WIP) and `Done` (you may have different names for the columns, but those three make up the basic concept). These three columns contains everything that is _in scope_. The attention span of this scope usually being only a few days. We don't need longer perspective in kanban, because we revisit the plan every single day. 

- The **`ToDo`** column is where team members pick up their next task
- The **`Work In Progress`** column is the tasks that are currently being worked on in the team
- The **`Done`** column are tasks that are completed

There are two other categories which together hold the (potential) tasks that are not - currently - in scope. _the backlog_ and _the inbox_.

- **The `Backlog`** holds everything that is _also_ important - just not in the span of the next couple of days perspective. It's tasks that the team know that they will have to do eventually, but not now. They will come into scope soon enough. During the daily revisit of the plan the team looks over the backlog - again - and see if there is anything there, that's ripe to bring into scope.

- **The `Inbox`** is any (potential) task this is neither in scope nor in the backlog. Typically a team will have several new issues being proposed fra various sources every day. Most of them comes from the team members themselves. As they work, new perspective or new problems surface. Or the new issues may come from sponsor, scrum master, end-users etc. These new proposed issues will be registered immediately as they surface. But the individual team member is often not in a position where it's possible to decide if a new issues has priority over existing ones; Should it got straight to the `ToDo` column? And if so, with what priority? Or should it go to the `Backlog`? Or is it even important? Perhaps it should be deleted straight away? Maybe it's a duplicate of something already registered in the plan? So consequently; it goes to the `Inbox`. The team must have a process for emptying the inbox at least once per day.

<img width="212" alt="image" align="left" src="https://github.com/thetechcollective/public-discussions/assets/155492/827b0df2-0ca7-4925-9125-eb088c3727b1"> 

## Real-life implementations

In Github work items are defined as _issues_. Issues belong to a specific repository and _projects_ are used to combine issues from multiple repositories into the same board or list view.

I implement the **`Inbox`** simply by _not_ setting the `status` field (the `status` field defines the columns in board view). This approach exploits the feature that in a board view in a GitHub project, you will automatically get an extra column most to the left with the title "No status" if there are any issues with the status field _not_ set. And on the contrary; if all issues do have a status then the "no status" column will not show at all. So the process of _emptying the inbox_ simply implies giving the issues without a status ...a status.

The **`backlog`** I implement by defining it as a new value in the `Status` field. But since I actually don't consider it part of my kanban, I will setup a separate board view I name "Kanban" in which I will deliberately _hide_ the `Backlog` column and then another board view I call "Backlog" or sometimes "Project grooming" in which I only see two columns; the `Backlog` and the `ToDo` column,  and then of course the automatic `No status` column if there are any. The "Kanban" view is then the every-day-team view, and the "Backlog" view is the one I uses to groom the project and prepare the plan - at least once every day.

# 👍 2: Keep everything in your `ToDo` list prioritized.

It's really not necessary - not even recommended - to prioritize anything other than the items in the **`ToDo`** column. That is; The `Inbox` and the `Backlog` don't need prioritizations - yet. The `ToDo` column is the entry (leftmost) in the kanban. Items move from left to right in the kanban board - consequently, the columns to the right of `ToDo` will inherit the priority you set when you introduce items to the kanban in `ToDo`.

But the kanban approach completely falls apart if the items in the `ToDo` column aren't prioritized. Kanban work like this: When a team member is done with a task, they will take a new one from the `ToDo` column. If items here aren't prioritized, there's a genuine risk, that the team member may pick up an item, that's not as important as another one that sits back untouched.

On the other hand; prioritizing items too early, before they enter the kanban, while they are still in either `Inbox` or `Backlog` may give a false indication that these items are expected to be included in the flow. I usually restrain myself from setting the proprity before I enter them into `ToDo`. But sometimes I need to have _some_ prioritization on the `backlog` as well. It may grow to a size, where it's huge, but my attention is on the short term 20-30 items that are coming up soon. For this kind of focus I utilize the board view's build-in order; I can drag and organize all items relative position to each other, and I use this feature, to pull the items that need my attention to the top of the column. The items relative positions are recorder behind the scene by GitHub project, and every time I come back to any board view, they have the same position as when I left them.

<img width="230" align="left" alt="image" src="https://github.com/thetechcollective/public-discussions/assets/155492/b09a8bc8-99c9-4744-873a-74928f645e2b">

### Real-life implementations

I've found that the MoSCoW [^moscow] prioritization method, in general works fine for most situations, and especially when working with iterative time-boxing techniques. so that's my favorite goto approach.

[^moscow]: A good short introduction to MoSCoW is the [Wikipedia entry on the topic](https://en.wikipedia.org/wiki/MoSCoW_method)

In GitHub Projects I usually implement a new single-select field called "Priority" or "MoSCoW", and then I define the four levels. I also implement a fifth level, which is not strictly included in the MoSCoW approach, but equally important in the processes of grooming - and that is `Duplicate`. It turns out that in your prioritization and your grooming process you often want to discard an issue - not because it's not appropriate (`Won't Have`) but simply because it's already being tracked elsewhere, and you want to keep things crisp and simple. When I use the priority `Duplicate` I make it a habit to _always_ give a reference to the issue it's a duplicate of. This link is automatically [also applied bi-directional](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls) on the _master_ issue. Neat and simple.

<br clear="both"/>

# 👍 3: Kanban is a _pull_ model - not a _push_ model. 

[The Toyota Way](https://www.tech-that.club/the-toyota-way/), Lean and consequently also agile all refers to a core principle: "Minimize Work In Progress". It's the core concept in the fabled "Just In Time" principle so central to The Toyota Way's _one-piece-flow_ that forces problems to surface early and hereby lay the foundation for _kaizen_ which means _continuous improvement_. 
 
Kanban means "signal" and the signal defines the _pull_. When a signal is raised, the pull starts.

In task management such as GitHub Project board, the pull comes from the column `In Progress`. As mentioned earlier; when a team member is done with a task and they clear it from the `In Progress` column; _that_ is when they signal that they are ready to take an another task. 

Long story short: Kanban is a simple, but central principle. In my experience we're now at the core, of why so many agile and lean processes seem to not deliver what they promised: The teams don't pay enough attention to the _pull_ and _Minimize WIP_ principle and consequently they allow their process to deteriorate into a mere, classic push model[^waste]. They talk the talk, but they don't walk the walk; They use the contemporary lingo of the context; _kanban_, but they don't understand or follow the principles.

[^waste]: [The Toyota Way](https://www.tech-that.club/the-toyota-way/) identifies three types of waste; [_muda_, _mura_ and _muri_](https://www.techtarget.com/whatis/definition/Muda-Mura-and-Muri). While _muda_ is the more commonly known type of waste; The kind of work, that does not add value, _muda_ is _unevenness_ which makes it really hard to plan and optimize and _muri_ is _overburden_ which is straining the process and causing bottlenecks or break downs. In lean inspired task management, such as kanban, the ideal is to have a steady even inflow of important work, which does not overburden team members. The _pull_ and the _Minimize WIP_ principles seek to optimise _muda_ and _muri_ respectively.
 
<img width="250" align="left" alt="image" src="https://github.com/thetechcollective/public-discussions/assets/155492/30b45f72-d040-4536-a7fe-4b84b3469f14">

### Real-life implementations

I would recommend the team to include this question in their frequent retrospectives:

> _"How are we doing, in terms of pull versus push? and in Minimizing Work In Progress?"_

\- _"Not doing all that well?"_ These are the first early symptoms that your process needs some more ❤️.

If a team member is really not supposed to work on more than one task at a time, it implies, that in a team of five, there can then only be maximum five issues in the "In Progress" column at the same time. GitHub Projects allows you to set that as a hard limit. Until a team is so experienced that all team members are capable of administer their individual WIP themselve, I recommend setting this limit, to the number of team members. It's like playing "musical chairs" or attending a "fishbowl discussion"; You need to find a vacant spot to participate.

I've seen and heard others who argue that the WIP limit should more conveniently be _two times the number of team members_, from the argument that sometimes a task gets stuck; Maybe you asked a question to someone and now you are stuck until you get a reply. So you are allowed to take on another issue instead.

I strongly disagree. Not that you shouldn't take on another issue while you wait; of course you shall. But you would have to clear the issue that is awaiting a reply out of the `In Progress` column while you wait.

\- _"..and take it where to?"_ 

Think of it; This is another kanban! When you receive the reply you are waiting for, that is literally a _signal_ that must cause a _pull_ ...back into the `In Progress` column.

In GitHub projects, I model this by adding an additional status `Awaiting reply` and then give the issue that new status and assign the issue to the person I'm waiting for. In GitHub Project's board view an additional status field value will automatically show up as an additional column in the board. I place all these additional _kanbans_ (_signals_ that triggers a _pull_) between `ToDo` and `In Progress`. Since they are now to then to the right of `ToDo` the logic follows that they are already prioritized, and that priority still applies. The only column that should be to the right of `In Progress` -is `Done´.

Consequently _sometimes_ things are not always moving left to right in the kanban. _Sometimes_ there is a set back where an item may take a step to the left. But still; items only _progresses_ to the right.

For a kanban status like `Awaiting reply` you can consider if you also need a date field to accompany it. Indicating how long you are willing to wait for that kanban before the issue needs escalation. Personally I most often do _not_ add dates like this to our issues. I just groom the plan every day, and I exploit the feature mentioned before, that in a board view I can order items relative to each other in a column, so I keep the one that needs most care and attention on top. 

You will probably experience the need for other kanbans (_statuses_, _signals_) like; "Needs a review", "Needs more research", "Time estimate is off" and if any of these needs to be monitored as separate statuses then I would model them all in the same way as the `Awaiting reply`. But practically, for the three examples just mentioned, I would probably deem them all to be mere variants of `Awaiting reply` and I would probably just use that.

**Note on assigning issues:** By this approach it's important to understand that we're putting important semantic meaning into _who_ is actually _assigned_ to what. So be meticulous about this. It's important that team members can easily see where their help is needed. They do this by means of paying attention to [notifications that they are assigned to items](https://github.com/notifications?query=reason%3Aassign) and by monitoring all [issues they are assigned to](https://github.com/issues/assigned). So be respectful about who you assign to what. If team members are expected to react on notification - they should not be spammed. Still you might need someones perspective or attention, You get that from [mentioning people or teams](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#mentioning-people-and-teams) in the comments.

# 👍 4: Groom your kanban at least once every day

**Consider the following - quite typical - anti-pattern scenario:**

> A team member is done with a task, and he then goes to the `ToDo` column to pick up a new one. From reading the titles most of the issues appear to be dealing with domains and competences which aren't exactly where he claim himself to be most proficient or efficient in, so it's hard do find one that's obviously well suited. 

> He then starts to read up on the issues one by one, but most of them has no further descriptions; they are all essentially just _titles_. Some are prioritized, but not all, and some are already assigned to other team members, kinda indicating that they are off limits to him, but waiting explicitly for these particular people. 

>But luckily he finds other tasks on the backlog which are far better suited for him. Actually he finds a handful of interesting and fun tasks, that he put there herself, so even if these too only consists of a title and no elaborate description - at least then he knows how to work them. 

> So he assigns himself to one of these backlogs issues and moves it straight to `Work in progress` and get cracking at it.

When kanban is done this way it implies:

- Tasks are executed in arbitrary order.
- _No one_ communicates with _anybody_.
- Team members only work with competences they already possess.
- It's not clear how solving a task generates value.
- _Mura_ and _muri_ - _Unevenness_ and _overburden_ is part of the game: Some are too busy, others have nothing sensible to do.

**OK, now consider this - ideal - scenario:**

> A pair of two team members finishes a task they have worked together following the principles of _paired programming_, despite the fact that it wasn't even strictly speaking a programming task. Now they are done and they go to the `ToDo` column and filter the view so they only see `Must Haves` and then sort ascending by `story point` so the _easiest_ low-hanging-fruit ones are at the top. Without any further investigation or reflection they take the one with the fewest story points, assigns themselves to the issue and moves it to `In Progress`. Only then do they open it to read what it's about. 

> One of them starts to laugh, the other gets all pale and anxious. The description on the issue is clear enough, concise and clearly described under the headings "Problem", "Solution" and "Value". But this is definitely not their field of expertise. They are looking at a learning curve as steep as Mont Blanc.

> First thing the do is to assign a third of the their team mates to the issue. She's the team champion in this particular field. They know well enough, that she's busy on another task, but together with the co-assignment to her, they leave a note for her in the comments on the issue to explain why they assigned her: "We need you help on this one 🙏". Then then start to elaborate a bit on the description of the "Problem" - the "Solution" suggested is way too technical to them, but the problem is clear enough, and they start investigating if there's another approach to this. They find a blogpost that describes what they believe to be the same problem, but a completely different solution. The add the link the blogpost to the comments and describe in just one single sentence the new approach they are taking. 

> After 15 mins their teamvmate adds a link in the comments to a tutorial on YouTube that she recommends to overcome the Mont Blanc'y learning curve. She also lets them know, that she hasn't time to join or take over the issue just now, but she validates that they are on the right track with their alternative solution approach and that she will stay assigned to the issue so she will automatically get notifications on every singe update in the comments. 

> The pair finishes the issue to their best effort 3 hours later. They spent way more time than indicated by the story point, but they are still proud about themselves. They learned a bunch. They give it the status "Needs a review". They unassign themselves and their champion team mate from the issue, but they do leave her a comment where they mention her specifically, indicating that they would love _her_ to review it. 

> Then they return to their filtered view on the `ToDo` column and pick the next most-important, low-hanging fruit.

When kanban is done this way it implies:

- Learning on the job.
- A clear indication of what the _problem_ is, what a _solution_ could look like, and the _value_ it would bring.
- Room for different thinking and alternative solutions to common problems.
- No stress from dead-lines, room for working in pairs, but still clear enough indications on expected complexity.
- A clear worklog, that serves both to enables asynchronous dialog and easy onboarding of team mates while being worked on and to ensure documentation of the solution.

The argument for grooming daily is, that it's simply required for the ideal scenario to have a chance over the anti-pattern scenario. 

👉 Items in `ToDo` must have a _priority_
👉 Items in `ToDo` must have an _measure on complexity_ (story point)
👉 Items in `ToDo` must have an unambiguous description or _problem_, _solution_ and _value_
👉 Avoid pre-assigning team members to issues that are not yet `In Progress`
👉 Items that are waiting for a _kanban_ must be monitored in case they need escalation
👉 `Inbox` must be monitored and emptied regularly, at least every day
👉 For task with same priority, less complexity is favored. 
👉 Breaking down task to contain less complexity avoids _mura_; _unevenness_

<img width="250" align="left" alt="image" src="https://github.com/thetechcollective/public-discussions/assets/155492/4386674b-87c6-43ef-9625-425c9ef2e3f2">

### Real-life implementations

I'm a big favor of the whole `#NoEstimates` paradigm, which abandons the idea that we can predict how long time a task will take, especially in a kanban flow that favors elimination of _muda_ - _unevenness_ by encouraging that everyone picks up the next - most important - task, even despite they might not have the highest proficiency in the field. A time estimate very much depends on who picks up the task - and we don't know that in advance; If we allowed task to be pre-assigned to specific individuals we would cause both _muda_ and probably also _muri_ - _overburden_ on the usual suspects.

\- _"OK So `#NoEstimate` it is then - now what?"_

Inspiration comes from the _XP - Extreem Programming_ approach which introduced _story points_. They were meant to be a measure of anticipated complexity in the _user stories_ as opposed to time estimate.

I like the concept, but the `#NoEstimate` paradigm suggests that if we want to measure rather than predict. then we should indicate _uncertainty_ rather than _complexity_. Complexity is not necessarily something that even represents a problem, if the team is very skilled, and eats complexity for breakfast. And also the _story point_ approach suggests that we take a well-known story, which everyone can relate to and knows the complexity off - and then all other stories must be assessed relative to that _foundation_  user story. Hmmm - not very practical, and still subject to speculations.

But _uncertainty_ is actually a measure of _not knowing_, _not having experience_, _a need for learning or investigation_ and again; this too depends on who the task is assigned to, so let's do this relative to the current team - not individuals.

Admittedly, this is something that probably requires the team to wiggle or wing it into being useful, but give it a go - and adjust; continuously improve.

In GitHub Projects I use a single-select field called `Uncertainty` (or stick with `complexity` if the semantical destination is too hippie) which uses the Fibonacci numbers up to and including 13; `1`, `2`, `3`, `5`, `8` and `13`. To set this measure imagine answering the following question:

> _"Given the combined, knowledge and skills af the current team, how is uncertainty assessed?"_

This is a measure that potentially can not be set by the project lead alone, it may require input from the entire team. Using the Fibonacci numbers enables team members to take a quick round of _planning poker_ which is a good way of facilitating a _kaizen blitz_ - and to make quick decisions - even when the debate gets heated. If need be, you can do this immediately after the daily standup.

Needless to say, this requires that "Problem", "Solution" and "Value" are described on the issue. So give it a quick draft. You will experience that very often the different hands played during _planning poker_ reflects that the details are ambiguous - it leaves room for opinions. Apparently we're not measuring but more likely steering by gut feelings. Cool, no problem; the discussions will help to clarify that - and the team will discuss and settle on common ground - and enable us to measure. The description will of course be updated accordingly.

Agree on what the Fibonacci numbers means - I argue, that even if we are not doing estimates, we are indeed doing _time-boxes_ - or _spikes_ as scrum likes to call them. And in that context it's OK, to see the Fibonacci numbers represent _hours_. As in 

> _"How much time would you set aside - literally, as a booking in your calendar - before you would expect some kind of valuable outcome from the endeavour?"_

It could be:

1️⃣ Open-and-shut case
2️⃣ Quickie - all is known
3️⃣ Research required
5️⃣ Experiments required
8️⃣ Training required 

By allowing numbers to represent hours we get to say, that any number higher than 8 indicates that it would take more than a full concentrated work day to generate a valuable outcome. And _every_ team member deserves the ability to close at least one task per day. So consequently any number higher than eight is too big. It requires the task to be broken down further, and then re-assessed.

<img width="300" align="left" alt="image" src="https://github.com/thetechcollective/public-discussions/assets/155492/1e79ff2c-2f26-487e-a9a3-fd4f2b141ff4">

The team should have _someone_[^teamlead] - a _team leader_ - appointed  to enable the team to go faster. The team lead is not a boss, but rather a janitor. A person who keeps the place tidy and neat, easy for everyone else to manoeuvre in. Constantly monitoring flow and re-enables it if there are any clots.

My recommendation is that the team does a standup every morning, but instead of making it a dull status meeting over information that is already logged - and asynchronously available to everyone - in the issues. Use the daily standup to assess team health; Reflect on _yesterday_, _today_ and _tomorrow_ and help eliminate waste, reflect on _muda_, _mura_ and _muri_. Maybe do a _sailboat_ reflection. Team members must support the team lead, by mentioning everything that needs support.

The team lead will facilitate the morning standup, make sure it's short and valuable to everyone, record actions that needs to be taken and if necessary follow up individually immediately after the standup. The team lead knows where everyone is. If someone is ill or for other reasons not working today, then team lead must be notified and adjust the limit on `Work In Progress` accordingly. The team lead will empty the task `Inbox` regularly - at least daily. The team lead will curate then task `backlog` regularly - at least daily - and bring task from there into scope if necessary. The team lead will pay constant attention to notifications from other team members and step in and help when needed _(leadership by presence)_. 

[^teamlead]: I call the _someone_ a _team lead_, but others would say _scrum master_, _project manager_ ...

**Being a team lead is role, not an expertise, the team lead may very well need help to fulfil this role. It's is the responsibility of the _group leader_ to offer support, mentorship and help to the team lead.**

**Comment by ravvnen** [comment]
@lakruzz - Should we do stand-up meetings these days in regards to dev-container progress?

**Reply by lakruzz** [reply]
I'm all for that:

> My recommendation is that the team does a standup every morning, but instead of making it a dull status meeting over information that is already logged - and asynchronously available to everyone - in the issues. Use the daily standup to assess team health; Reflect on yesterday, today and tomorrow and help eliminate waste, reflect on muda, mura and muri. Maybe do a sailboat reflection. Team members must support the team lead, by mentioning everything that needs support.

**Reply by ravvnen** [reply]
What time would suit you? (We could evt set a recurring time, for simplicity sake)
